man/man1/indomcachectl.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH INDOMCACHECTL 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3indomcachectl\f1 \- manage instance domain cache files
.SH SYNOPSIS
.B $PCP_BINADM_DIR/indomcachectl
[\fB\-w\fR]
\fB\-l\fR
.I indom
.sp
.B $PCP_BINADM_DIR/indomcachectl
[\fB\-w\fR]
[\fB\-m\fR \fImode\fR]
[\fB\-g\fR \fIgroup\fR]
[\fB\-u\fR \fIuser\fR]
.I indom
.SH DESCRIPTION
.de EX
.in +0.5i
.ie t .ft CB
.el .ft B
.ie t .sp .5v
.el .sp
.ta \\w' 'u*8
.nf
..
.de EE
.fi
.ie t .sp .5v
.el .sp
.ft R
.in
..
.B indomcachectl
assists in the management of the Instance Domain cache files that
are stored in
.BR $PCP_VAR_DIR/config/pmda .
.PP
The mandatory
.I indom
parameter uses a 2-dotted format to specify the
.I domain
and
.I serial
fields of the instance domain number.
The required values can be reported for a particular metric using the
.B \-d
option to
.BR pminfo (1).
.PP
The list of available options can be retrieved using the
.BR \-? / \-\-help
option, which displays a usage message and exits.
.PP
If the
.BR \-l / \-\-list
option is specified the contents of the instance domain cache file
are dumped on
.I stdout
and
.I indomcachectl
exits.
.PP
Otherwise an empty instance domain cache file will be created.
The user ownership of the created file is assigned to the uid
of the caller, but may optionally assigned to the uid of
.I user
from the
.BR passwd (5)
file if the
.BR \-u / \-\-user
option is specified.
The group ownership of the created file is assigned to the gid
of the caller, else the default gid of the user
.I user
if the
.I \-u
option is specified, else the gid of
.I group
from the
.BR group (5)
file if the
.BR \-g / \-\-group
option is specified.
.PP
The instance domain cache file is created with mode 0660 by
default (some PMDAs use group read-write permission to access
and update the contents of the file)
\- if this
does not suit the
.BR \-m / \-\-mode
option may be used to set the mode to the (octal) value
.IR mode .
.PP
Because the indom cache files are precious to the associated PMDA
.I indomcachectl
will not over-write an existing indom cache file.
.PP
Operation is usually silent, except for errors (or warnings if the
.BR \-w / \-\-warning
option is used).
The exit status is 0 for success, 1 for failure.
.SH OTHER MANAGEMENT OPTIONS
The files in
.BR $PCP_VAR_DIR/config/pmda
are all text files (one per cached instance domain)
so judicious surgery may be used to make changes
or even remove the files
in extreme cases
.B provided
the PMDA is not executing.
.PP
Here is an example for the instance domain number
60.17 that is associated with the
network.interface.* metrics.
.EX
$ cat /var/lib/pcp/config/pmda/60.17
2 0 2147483647
0 1712727825 ens33
1 1712727825 virbr0
2 1712727825 lo
.EE
.PP
The first line is a magic header, do not change it.
Following that is one line per instance with the instance identifier
(an integer) in the first field, some magic in the second field
(do not change this) and the external instance name in the third field.
.SH SEE ALSO
.BR PCPIntro (1),
.BR chmod (1),
.BR pminfo (1),
.BR PMAPI (3),
.BR PMDA (3),
.BR pmdaCache (3),
.BR group (5)
and
.BR passwd (5).

.\" control lines for scripts/man-spell
.\" +ok+ virbr ens sp {from network interfaces}


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/indomcachectl/indomcachectl.c
/*
 * Copyright (c) 2024 Ken McDonell.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * Debug flags
 * <none>
 */

#include "pmapi.h"
#include "pmda.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <pwd.h>
#include <grp.h>

static int
allmyoptions(int opt, pmOptions *optsp)
{
    /* they all belong to me */
    return 1;
}

static pmLongOptions longopts[] = {
    PMAPI_OPTIONS_HEADER("Options"),
    { "group", 1, 'g', "GROUP", "create cache file owned by group GROUP" },
    { "list", 0, 'l', NULL, "list indom cache contents" },
    { "mode", 1, 'm', "MODE", "create cache file with mode MODE (octal)" },
    { "user", 1, 'u', "USER", "create cache file owned user USER" },
    { "warning", 0, 'w', NULL, "issue warnings" },
    PMOPT_HELP,
    PMAPI_OPTIONS_END
};

static pmOptions opts = {
    .short_options = "g:lm:u:w?",
    .long_options = longopts,
    .short_usage = "[options] domain.serial",
    .override = allmyoptions,
};

int
main(int argc, char **argv)
{
    int		lflag = 0;
    int		wflag = 0;
    int		c;
    int		domain, serial;
    int		sts;
    int		badarg = 0;
    pmInDom	indom;
    char	*user = NULL;
    uid_t	uid = geteuid();
    struct passwd	*pwp = NULL;
    char	*group = NULL;
    gid_t	gid = getegid();
    struct group	*grp = NULL;
    int		mflag = 0;
    mode_t	mode = 0660;	/* default user:rw group:rw other:none */
    char	*endp;
    long	tmp_mode;
    int		sep = pmPathSeparator();
    char	pathname[MAXPATHLEN];

    while ((c = pmGetOptions(argc, argv, &opts)) != EOF) {
	switch (c) {

	    case 'g':		/* group */
		grp = getgrnam(opts.optarg);
		if (grp == NULL) {
		    fprintf(stderr, "%s: Error: getgrnam(%s) failed\n", pmGetProgname(), opts.optarg);
		    badarg++;
		}
		else {
		    group = opts.optarg;
		    gid = grp->gr_gid;
		}
		break;

	    case 'l':		/* list/dump */
		lflag++;
		break;

	    case 'm':		/* mode */
		tmp_mode = strtol(opts.optarg, &endp, 8);
		if (*endp != '\0' || tmp_mode > 0777) {
		    fprintf(stderr, "%s: Error: mode (%s) is not a valid\n", pmGetProgname(), opts.optarg);
		    badarg++;
		}
		else {
		    mflag++;
		    mode = (mode_t)tmp_mode;
		}
		break;

	    case 'u':		/* user */
		pwp = getpwnam(opts.optarg);
		if (pwp == NULL) {
		    fprintf(stderr, "%s: Error: getpwnam(%s) failed\n", pmGetProgname(), opts.optarg);
		    badarg++;
		}
		else {
		    user = opts.optarg;
		    uid = pwp->pw_uid;
		}
		break;

	    case 'w':		/* warnings */
		wflag++;
		break;
	}

    }

    if (badarg)
	exit(1);

    if (opts.errors || (opts.flags & PM_OPTFLAG_EXIT) || opts.optind != argc-1) {
	int		exitsts;
	exitsts = !(opts.flags & PM_OPTFLAG_EXIT);
	pmUsageMessage(&opts);
	exit(exitsts);
    }

    if ((user != NULL || group != NULL || mflag != 0) && lflag) {
	fprintf(stderr, "%s: Error: -l is mutually exclusive with -u, -g and -m\n", pmGetProgname());
	exit(1);
    }

    if (user != NULL && group == NULL) {
	/* -u and no -g, use user's default group */
	gid = pwp->pw_gid;
    }

    if (sscanf(argv[opts.optind], "%u.%u", &domain, &serial) != 2) {
	fprintf(stderr, "%s: Error: bad format domain.serial: %s\n", pmGetProgname(), argv[opts.optind]);
	exit(1);
    }
    /*
     * domain is limited to 9 bits, serial to 22 bits ... both need
     * to be positive
     */
    if (domain < 0 || domain >= (1<<9) || serial < 0 || serial >= (1<<22)) {
	fprintf(stderr, "%s: Error: out-of-range domain.serial: %s\n", pmGetProgname(), argv[opts.optind]);
	exit(1);
    }
    indom = pmInDom_build(domain, serial);

    /*
     * try loading the cache ... needed for -l, doesn't matter otherwise
     */
    if ((sts = pmdaCacheOp(indom, PMDA_CACHE_LOAD)) < 0) {
	if (wflag)
	    fprintf(stderr, "Warning: %s: PMDA_CACHE_LOAD: %s\n", pmInDomStr(indom), pmErrStr(sts));
    }
    else if (!lflag) {
	/* don't clobber an existing indom cache file */
	fprintf(stderr, "%s: Error: %s: indom cache file already exists\n", pmGetProgname(), pmInDomStr(indom));
	exit(1);
    }

    if (lflag) {
	int	inst;
	char	*iname;
	int	numinst = 0;
	if (sts < 0) {
	    fprintf(stderr, "%s: Error: %s: failed to load indom cache file\n", pmGetProgname(), pmInDomStr(indom));
	    exit(1);
	}
	if ((sts = pmdaCacheOp(indom, PMDA_CACHE_ACTIVE)) < 0) {
	    if (wflag)
		fprintf(stderr, "Warning: %s: PMDA_CACHE_ACTIVE: %s\n", pmInDomStr(indom), pmErrStr(sts));
	}
	if ((sts = pmdaCacheOp(indom, PMDA_CACHE_WALK_REWIND)) < 0) {
	    if (wflag)
		fprintf(stderr, "Warning: %s: PMDA_CACHE_WALK_REWIND: %s\n", pmInDomStr(indom), pmErrStr(sts));
	}
	printf("Instance domain %s cache contents ...\n", pmInDomStr(indom));
	while ((inst = pmdaCacheOp(indom, PMDA_CACHE_WALK_NEXT)) != -1) {
	    if ((sts = pmdaCacheLookup(indom, inst, &iname, NULL)) < 0) {
		if (wflag) {
		    fprintf(stderr, "Warning: %s: Lookup failed for inst=%d: %s\n", pmInDomStr(indom), inst, pmErrStr(sts));
		}
		continue;
	    }
	    if (numinst == 0)
		printf("%8.8s %s\n", "inst_id", "inst_name");
	    printf("%8d %s\n", inst, iname);
	    numinst++;
	}
	if (numinst == 0)
	    printf("No entries\n");

	exit(0);
    }

    /*
     * initialize ...
     */
    pmsprintf(pathname, sizeof(pathname), "%s%cconfig%cpmda%c%s",
	pmGetOptionalConfig("PCP_VAR_DIR"), sep, sep, sep, pmInDomStr(indom));
    if ((sts = pmdaCacheOp(indom, PMDA_CACHE_WRITE)) < 0) {
	fprintf(stderr, "Error: %s: create %s failed: %s\n", pmGetProgname(), pathname, pmErrStr(sts));
	exit(1);
    }
    if (user || group) {
	/* change owner and group */
	if (chown(pathname, uid, gid) < 0) {
	    fprintf(stderr, "%s: Error: created %s, but cannot change ownership to uid=%d gid=%d: %s\n", pmGetProgname(), pathname, uid, gid, pmErrStr(-oserror()));
	    exit(1);
	}
    }
    /* explicitly set mode ... */
    if (chmod(pathname, mode) < 0) {
	fprintf(stderr, "%s: Error: created %s, but cannot change mode to 0%o: %s\n", pmGetProgname(), pathname, mode, pmErrStr(-oserror()));
	exit(1);
    }

    exit(0);
}
